### PR TITLE
Refactor and benchmark

### PR DESCRIPTION
**Pull request by vwxyzjn**
This PR does a few refactor. Some raw thoughts:

* For better experiment tracking, we should also create a few more variables
    * `query_dataset="imdb"`
    * `reward_model="sentiment-analysis:lvwerra/distilbert-imdb"`, it could work with a pipeline or a trained reward model
* By default, we probably should use a vanilla model like `gpt2` in place of `lvwerra/gpt2-imdb`
* By default, we should demonstrate end-to-end training (involving a reward model training then policy training).

We have multiple benchmark axes:

- w/ different models (gpt2, gpt2-xl, falcon, llama2)
    - key research engineering questions
        - how do different model sizes scale?
        - **given that the preference labels come from a source model `M_s` (e.g., gpt2), how does that affect the performance of a target model `M_t` (e.g., falcon, gptj, llama2)?**
            - This is actually an important assumption we have been operating.
- w/ and w/o gradient accumulation / multi-GPU
    - key research engineering question: do we need to whiten advantage across the entire batch?
- w/ and w/o peft
    - key research engineering question: how well does PEFT work with RL
- w/ and w/o quantization or 4 bits
    - key research engineering question: how well does quantization work with RL training
- w/ and w/o deepspeed
    - sanity check to make sure it works.
- w/ different datasets
    - TRL’s typical imdb sentiment
    - OAI’s sentiment dataset (https://github.com/openai/lm-human-preferences)
    - summarize from feedback ( https://github.com/openai/summarize-from-feedback)
    - helpfulness vs harmlessness (https://huggingface.co/datasets/Anthropic/hh-rlhf)

We can probably have a `train.py` that can do 

```
accelerate launch train.py --config deepspeed.yaml \
    --model_name falcon-40b \
    --query_dataset book \
    --label_dataset openai-sentiment-label \
    --gradient_accumulation_steps=12 \
```

## Uses `tyro` to eliminate duplicate code / comments. 

Help text also works

image

<table>
<tr>
<td> tyro </td> <td> existing </td>
</tr>
<tr>
<td>

```python
@dataclass
class ScriptArguments:
    ppo: PPOConfig = field(
        default_factory=lambda: PPOConfig(
            model_name="lvwerra/gpt2-imdb",
            learning_rate=1.41e-5,
            log_with=None,
            mini_batch_size=128,
            batch_size=128,
            gradient_accumulation_steps=1,
            early_stopping=False,
            target_kl=6,
            kl_penalty="kl",
            seed=0,
        )
    )
args = tyro.cli(ScriptArguments)

print(args.ppo.seed)
```

</td>
<td>
    
```python
@dataclass
class ScriptArguments:
    """
    The name of the Casual LM model we wish to fine with PPO
    """

    # NOTE: gpt2 models use Conv1D instead of Linear layers which are not yet supported in 8 bit mode
    # models like gpt-neo* models are more suitable.
    model_name: Optional[str] = field(default="lvwerra/gpt2-imdb", metadata={"help": "the model name"})
    log_with: Optional[str] = field(default=None, metadata={"help": "use 'wandb' to log with wandb"})
    learning_rate: Optional[float] = field(default=1.41e-5, metadata={"help": "the learning rate"})
    mini_batch_size: Optional[int] = field(default=128, metadata={"help": "the PPO minibatch size"})
    batch_size: Optional[int] = field(default=128, metadata={"help": "the batch size"})
    gradient_accumulation_steps: Optional[int] = field(
        default=1, metadata={"help": "the number of gradient accumulation steps"}
    )
    early_stopping: Optional[bool] = field(default=False, metadata={"help": "whether to early stop"})
    use_peft: Optional[bool] = field(default=False, metadata={"help": "whether to use peft"})
    use_seq2seq: Optional[bool] = field(default=False, metadata={"help": "whether to use seq2seq models"})
    kl_penalty: Optional[str] = field(
        default="kl",
        metadata={
            "help": "kl penalty options: 'kl': model_logp - ref_logp,  'abs': abs(kl),  'mse': mean squared error mse(kl) and 'full': the actual kl for all tokens in the distribution"
        },
    )
    target_kl: Optional[float] = field(default=0.1, metadata={"help": "kl target for early stopping"})
    seed: Optional[int] = field(default=0, metadata={"help": "the random seed"})
    use_score_scaling: Optional[bool] = field(default=False, metadata={"help": "Use score scaling"})
    use_score_norm: Optional[bool] = field(
        default=False, metadata={"help": "Use score normalization. Only applicable if use_score_scaling is True"}
    )
    score_clip: Optional[float] = field(default=None, metadata={"help": "Score clipping"})


parser = HfArgumentParser(ScriptArguments)
script_args = parser.parse_args_into_dataclasses()[0]

config = PPOConfig(
    model_name=script_args.model_name,
    learning_rate=script_args.learning_rate,
    log_with=script_args.log_with,
    mini_batch_size=script_args.mini_batch_size,
    batch_size=script_args.batch_size,
    gradient_accumulation_steps=script_args.gradient_accumulation_steps,
    early_stopping=script_args.early_stopping,
    target_kl=script_args.target_kl,
    kl_penalty=script_args.kl_penalty,
    seed=script_args.seed,
    use_score_scaling=script_args.use_score_scaling,
    use_score_norm=script_args.use_score_norm,
    score_clip=script_args.score_clip,
)
```
</td>
</tr>
</table>

## add accelerate logging

Log `global_backward_batch_size` `global_batch_size` `world_size`

### CODE REVIEW

File: benchmark/plot.sh [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# pip install openrlbenchmark==0.2.1a5
+PR_662_TAG=v0.4.7-55-g110e672
+PR_662_NAME=PR-662
+
+python -m openrlbenchmark.rlops_multi_metrics \
```
Comment (lvwerra) [inline benchmark/plot.sh:5]: I am very confused by how these work :D They don't seem to be using the same `benchmark/benchmark.py` right? Can you somewhere document what that command exactly does, otherwise it looks hard to maintain for anybody else.

**Review by lvwerra — COMMENTED** [review]
Hi @vwxyzjn, looks really good to me. The openbenchmark code itself is a bit obscure to me so if we could document well how it works that would be great. Left some comments here and there but in general happy to add it. Since this is not strictly a part of the library and more part of our test suite we can also be a bit more experimental here :)

Regarding `tyro` - this looks good to me in general, would also love to hear @younesbelkada feedback who maybe also knows about design decisions in `transformers` about the CLI.

File: benchmark/upload_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+from huggingface_hub import HfApi
+
+
+api = HfApi()
+
+api.upload_folder(
+    folder_path="benchmark/trl",
+    path_in_repo="images/benchmark",
+    repo_id="trl-internal-testing/example-images",
```
[the remark this answers is not in the document]
Comment (vwxyzjn) [reply benchmark/upload_benchmark.py:9]: For current baselines, we can probably just do this, but for future PRs, we should upload to the corresponding tag like https://github.com/vwxyzjn/test-ci/blob/314e2f1c4e8ea1af11f448ce856254fe5ae7b5c8/benchmark/plot.sh#L11-L15

File: trl/trainer/ppo_config.py [hunk 3]
```diff
@@ -59,116 +61,103 @@ def autotag() -> str:
...
+    use_score_norm: bool = False
+    """Use score normalization. Only applicable if use_score_scaling is True"""
+    score_clip: Optional[float] = None
+    """Score clipping"""
+
+    # Model detection
+    is_encoder_decoder: Optional[tyro.conf.Suppress[bool]] = None
+    """TO BE FILLED In RUNTIME: Whether the model is an encoder-decoder model"""
+    is_peft_model: Optional[tyro.conf.Suppress[bool]] = None
+    """TO BE FILLED In RUNTIME: Whether the model is a PEFT model"""
```
[the remark this answers is not in the document]
Comment (vwxyzjn) [reply trl/trainer/ppo_config.py:160]: `tyro.conf.Suppress[bool]` it's basically preventing `is_peft_model` from appearing in the help text. We don't want users to input these variables because they will be populated in the runtime.

File: examples/scripts/sentiment_tuning.py [hunk 4]
```diff
@@ -16,11 +16,12 @@
 from typing import Optional
 
 import torch
+import tyro
 from accelerate import Accelerator
 from datasets import load_dataset
 from peft import LoraConfig
 from tqdm import tqdm
-from transformers import AutoTokenizer, HfArgumentParser, pipeline
```
[the remark this answers is not in the document]
Comment (vwxyzjn) [reply examples/scripts/sentiment_tuning.py:23]: Absolutely. If we decide to go with `tyro,` I will follow up with PRs to change other examples.

File: benchmark/plot.sh (hunk 1 above)
Comment (vwxyzjn) [reply benchmark/plot.sh:5]: Nice catch! The documentation is https://github.com/openrlbenchmark/openrlbenchmark#get-started. Should I just put the link to the documentation at the top of `plot.sh`?

File: benchmark/plot.sh [hunk 5]
```diff
@@ -0,0 +1,50 @@
+# pip install openrlbenchmark==0.2.1a5
```
[the remark this answers is not in the document]
Comment (vwxyzjn) [reply benchmark/plot.sh:1]: Maybe we can do this along with the full CI PR like https://github.com/vwxyzjn/test-ci/blob/main/benchmark/plot.sh https://github.com/vwxyzjn/test-ci/blob/main/benchmark.sh? The PR as it stands just adds documentation. I could not chain something like `make full_run` with `make plot` because it's unclear what benchmarks have been run.

[the remark this answers is not in the document]
Comment (vwxyzjn) [reply benchmark/plot.sh:1]: Actually just added the benchmark scripts and updated makefile :)

File: benchmark/upload_benchmark.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vwxyzjn) [reply benchmark/upload_benchmark.py]: Done.

File: trl/trainer/ppo_config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lvwerra) [reply trl/trainer/ppo_config.py:160]: sounds good!

File: benchmark/upload_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
+from dataclasses import dataclass, field
+from typing import List
+
+import tyro
+from huggingface_hub import HfApi
+
+
+@dataclass
+class Args:
+    tags: List[str] = field(default_factory=lambda: ["no-tag-1-gb906f3d"])
```
Comment (lvwerra) [inline benchmark/upload_benchmark.py:10]: looks like the tag is not used?

File: benchmark/plot.sh (hunk 1 above)
Comment (lvwerra) [reply benchmark/plot.sh:6]: ok sounds good

**Review by younesbelkada — APPROVED** [review]
Thank a lot for this great work , as discussed offline :D !
Feel free to merge once the CI is green

File: examples/scripts/sentiment_tuning.py [hunk 7]
```diff
@@ -178,11 +154,12 @@ def collator(data):
...
     device = 0 if torch.cuda.is_available() else "cpu"  # to avoid a `pipeline` bug
 ds_plugin = ppo_trainer.accelerator.state.deepspeed_plugin
+task, model_name = args.ppo_config.reward_model.split(":")
 if ds_plugin is not None and ds_plugin.is_zero3_init_enabled():
     with ds_plugin.zero3_init_context_manager(enable=False):
-        sentiment_pipe = pipeline("sentiment-analysis", model=script_args.reward_model_name, device=device)
+        sentiment_pipe = pipeline(task, model=model_name, device=device)
 else:
-    sentiment_pipe = pipeline("sentiment-analysis", model=script_args.reward_model_name, device=device)
+    sentiment_pipe = pipeline(task, model=model_name, device=device)
```
Comment (vwxyzjn) [inline examples/scripts/sentiment_tuning.py:163]: Here we fixed the merge conflict and made a generalization for @lewtun's new `reward_model_name` arg.
